File: src/homing2.cpp
Rcpp::NumericMatrix start(startS);
Rcpp::NumericMatrix current(currentS);
double threshold = as<double>(thresholdS);
double alpha = as<double>(alphaS);

int n = start.nrow();

Rcpp::NumericMatrix disp(start.nrow(), start.ncol());
disp.fill(0.0);

for (int i = 0; i < n; i++) {
        
  double dx = current(i, 0) - start(i, 0);
  double dy = current(i, 1) - start(i, 1);
  double l = sqrt(pow(dx, 2) + pow(dy, 2));

  if (l > 1e-6){
    l = alpha * (l - threshold) / l;
    disp(i, 0) -= dx * l / 2;
    disp(i, 1) -= dy * l / 2;
  }
}

return disp;

File: R/homing2.r
homing2 <- function(start, current, threshold, alpha){
	n <- dim(start)[1]
	disp <- matrix(data = 0, nrow = n, ncol = 2)

	dx <- current[, 1] - start[, 1]
	dy <- current[, 2] - start[, 2]
	
	# springpp <- cxxfunction( signature(), paste( readLines( "spring.cpp"), collapse = "\n" ), plugin = "Rcpp")
	
	for (i in 1:n){
		l <- sqrt(dx[i]^2 + dy[i]^2)
		# one-sided spring
		# want l = 0 if within distance
		if (l > 1e-6){
			l <- alpha * (l - threshold) / l
			disp[i, 1] <- disp[i, 1] - dx[i] * l / 2
			disp[i, 2] <- disp[i, 2] - dy[i] * l / 2
		}
	}
	
	disp
}